tls/s2n_client_cert_verify.c

```diff
@@ -74,14 +74,16 @@ int s2n_client_cert_verify_send(struct s2n_connection *conn)
     GUARD(s2n_hash_copy(&conn->handshake.ccv_hash_copy, &hash_state));
 
     struct s2n_cert_chain_and_key *cert_chain_and_key = conn->handshake_params.our_chain_and_key;
-    struct s2n_blob signature = {0};
-    signature.size = s2n_pkey_size(cert_chain_and_key->private_key);
-    GUARD(s2n_stuffer_write_uint16(out, signature.size));
-    signature.data = s2n_stuffer_raw_write(out, signature.size);
-    notnull_check(signature.data);
+
+    DEFER_CLEANUP(struct s2n_blob signature = {0}, s2n_free);
+    uint32_t max_signature_size = s2n_pkey_size(cert_chain_and_key->private_key);
+    GUARD(s2n_alloc(&signature, max_signature_size));
 
     GUARD(s2n_pkey_sign(cert_chain_and_key->private_key, chosen_sig_scheme.sig_alg, &conn->handshake.ccv_hash_copy, &signature));
```

Comment:
Just a callout for a second opinion: I was looking at the code for s2n_ecdsa_sign and I'm having trouble wrapping my head around a few of the lines for signature size calculation. It doesn't look like the size check in here could ever be triggered given how we're allocating the blob size and how ECDSA_sign works. I think it should still work though as written.

https://github.com/awslabs/s2n/blob/d785051062037e9d48425f790ba5e5ce0026c5de/crypto/s2n_ecdsa.c#L56-L59

Reply:
I think line 58 is basically a sanity check. We never expect it to be triggered.

signature->size is set to the maximum size of the signature (via s2n_pkey_size). However, due to how DER encoding works, the actual size of the signature might be one or two bytes less. The actual size should never be more, because signature->size is already the maximum.

Reply:
Hypothetically signatures could be *much* shorter, but it happens very rarely.

25% of the time it will be the maximum size
~1/1024 of the time it will be 3 bytes less
Even smaller probabilities for smaller ones.
(Back of the envelope calcs for that.)